src/dune_rules/obj_dir.ml

```diff
@@ -293,27 +293,35 @@ module Module = struct
     let dir = cm_dir t kind visibility in
     relative t dir obj_name
 
-  let cm_file_unsafe t m ~kind =
-    let ext = Cm_kind.ext kind in
-    obj_file t m ~kind ~ext
+  let has_impl_if_needed m ~kind =
+    match (kind : Cm_kind.t) with
+    | Cmo
+    | Cmx ->
+      Module.has m ~ml_kind:Impl
+    | Cmi -> true
 
-  let o_file_unsafe t m ~ext_obj = obj_file t m ~kind:Cmx ~ext:ext_obj
+  let raise_no_impl m ~kind =
+    Code_error.raise "module has no implementation"
+      [ ("m", Module.to_dyn m); ("kind", Cm_kind.to_dyn kind) ]
+
+  let o_file_unsafe t m ~ext_obj =
+    let kind = Cm_kind.Cmx in
+    if Module.has m ~ml_kind:Impl then
+      obj_file t m ~kind ~ext:ext_obj
+    else
+      raise_no_impl m ~kind
 
   let cm_file t m ~(kind : Cm_kind.t) =
-    let has_impl = Module.has m ~ml_kind:Impl in
-    match kind with
-    | Cmx
-    | Cmo
-      when not has_impl ->
+    if has_impl_if_needed m ~kind then
+      let ext = Cm_kind.ext kind in
+      Some (obj_file t m ~kind ~ext)
+    else
       None
-    | _ -> Some (cm_file_unsafe t m ~kind)
 
-  let cm_public_file_unsafe t m ~kind =
-    let ext = Cm_kind.ext kind in
-    let base = cm_public_dir t kind in
-    let obj_name = Module.obj_name m in
-    let fname = Module_name.Unique.artifact_filename obj_name ~ext in
-    relative t base fname
+  let cm_file_unsafe t m ~kind =
```

Comment:
Since the function now makes the relevant check and raises if it's violated, we should call it `cm_file_exn`, not `cm_file_unsafe`.